internal/datastore/mysql/datastore.go

```diff
@@ -0,0 +1,648 @@
+		if err != nil {
+			return nil, fmt.Errorf(errUnableToQueryTuples, err)
+		}
+		defer migrations.LogOnError(ctx, rows.Close)
+
+		span.AddEvent("Query issued to database")
+
+		var tuples []*core.RelationTuple
+		for rows.Next() {
+			nextTuple := &core.RelationTuple{
+				ObjectAndRelation: &core.ObjectAndRelation{},
+				User: &core.User{
+					UserOneof: &core.User_Userset{
+						Userset: &core.ObjectAndRelation{},
+					},
+				},
+			}
+			userset := nextTuple.User.GetUserset()
+			err := rows.Scan(
+				&nextTuple.ObjectAndRelation.Namespace,
+				&nextTuple.ObjectAndRelation.ObjectId,
+				&nextTuple.ObjectAndRelation.Relation,
+				&userset.Namespace,
+				&userset.ObjectId,
+				&userset.Relation,
+			)
+			if err != nil {
+				return nil, fmt.Errorf(errUnableToQueryTuples, err)
+			}
+
+			tuples = append(tuples, nextTuple)
+		}
+		if err := rows.Err(); err != nil {
+			return nil, fmt.Errorf(errUnableToQueryTuples, err)
+		}
+		span.AddEvent("Tuples loaded", trace.WithAttributes(attribute.Int("tupleCount", len(tuples))))
+		return tuples, nil
+	}
+}
+
+type Datastore struct {
+	db            *sql.DB
+	driver        *migrations.MysqlDriver
+	querySplitter *common.TupleQuerySplitter
+	url           string
+
+	revisionFuzzingTimedelta time.Duration
+	gcWindowInverted         time.Duration
+	gcInterval               time.Duration
+	gcMaxOperationTime       time.Duration
+	watchBufferLength        uint16
+
+	gcGroup  *errgroup.Group
+	gcCtx    context.Context
+	cancelGc context.CancelFunc
+
+	createTxn     string
+	createBaseTxn string
+
+	*QueryBuilder
+}
+
+// Close closes the data store.
+func (mds *Datastore) Close() error {
+	// FIXME dupe from postgres datastore - need to refactor
+	mds.cancelGc()
+	if mds.gcGroup != nil {
+		if err := mds.gcGroup.Wait(); err != nil {
+			log.Error().Err(err).Msg("error waiting for garbage collector to shutdown")
+		}
+	}
+	return mds.db.Close()
+}
+
+// FIXME dupe from postgres datastore - need to refactor
+func (mds *Datastore) runGarbageCollector() error {
+	log.Info().Dur("interval", mds.gcInterval).Msg("garbage collection worker started for mysql driver")
+
+	for {
+		select {
+		case <-mds.gcCtx.Done():
+			log.Info().Msg("shutting down garbage collection worker for mysql driver")
+			return mds.gcCtx.Err()
+
+		case <-time.After(mds.gcInterval):
+			err := mds.collectGarbage()
+			if err != nil {
+				log.Warn().Err(err).Msg("error when attempting to perform garbage collection")
+			}
+		}
+	}
+}
+
+// FIXME dupe from postgres datastore - need to refactor
+func (mds *Datastore) getNow(ctx context.Context) (time.Time, error) {
+	// Retrieve the `now` time from the database.
+	nowSQL, nowArgs, err := getNow.ToSql()
+	if err != nil {
+		return time.Now(), err
+	}
+
+	var now time.Time
+	err = mds.db.QueryRowContext(datastore.SeparateContextWithTracing(ctx), nowSQL, nowArgs...).Scan(&now)
+	if err != nil {
+		return time.Now(), err
+	}
+
+	// RelationTupleTransaction is not timezone aware
```

Comment:
Is this the case in MySQL?

Reply:
This should theoretically not be necessary, but while learning on how MySQL did this, I found some behaviour I couldn't quite understand.
- MySQL is governed by a global timezone parameter.
- `DATETIME` is stored with timezone information.
- sessions can also have their own timezone and override the server global
- when we query the `RelationTupleTransaction`, the driver is able to obtain `DATETIME` honoring the timezone it was stored with. The `parseTime` flag in the go driver enables properly unmarshalling the query response into a `time.Time`
- However, when performing a query without a table like `SELECT NOW()`, somewhere in the way the timezone wasn't being recognized. I spent some time debugging the driver but couldn't find the answer. With some tweaks to the `TestTransactionTimestamps` I managed to reproduce the problem, and I fixed it by forcing the timezone in the query `SELECT UTC_TIMESTAMP()`.

I ended up leaving it this way because I found it easier to convert all `time.Time` instances flowing through SpiceDB to UTC just to make it easier to visually inspect it while debugging, so it's here for a different reason that the PSQL driver. I'll adjust the comment.

Reply:
addressed in https://github.com/authzed/spicedb/pull/525/commits/2c9fcd1cfee69ff8aa1bafe57979bea90dcb0216